File: InstagramLab2/app/src/main/java/com/example/savenews/PostsAdapter.java
package com.example.savenews;

import android.text.Html;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;
import java.util.List;


public class PostsAdapter extends RecyclerView.Adapter<PostsAdapter.NewsViewHolder> {

    List<Posts> main_list;
    private @Nullable ItemClickListener listener;
    private @Nullable FragmentButtonListener fragmentButtonListener;
    private @Nullable FragmentLikeListener fragmentLikeListener;

    public PostsAdapter(List<Posts> postsList,
                        @Nullable ItemClickListener listener,
                        @Nullable FragmentButtonListener fragmentButtonListener,
                        @Nullable FragmentLikeListener fragmentLikeListener) {
        Posts.postsList = postsList;
        main_list = postsList;
        this.listener = listener;
        this.fragmentButtonListener = fragmentButtonListener;
        this.fragmentLikeListener = fragmentLikeListener;

    }

    @NonNull
    @Override
    public NewsViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(
                parent.getContext())
                .inflate(
                        R.layout.item_news,
                        null,
                        false);
        RecyclerView.LayoutParams params = new RecyclerView.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
        );
        view.setLayoutParams(params);
        return new NewsViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull final NewsViewHolder holder, final int position) {
        final Posts posts = Posts.postsList.get(getItemViewType(position));
        holder.profilePhoto.setImageResource(posts.getProfilePhoto());
        holder.postImage.setImageResource(posts.getPostImage());
        holder.author.setText(posts.getAuthor());
        String s = "<b>"+ posts.getAuthor()+ "</b>" + " "+ posts.getPostText();
        holder.postText.setText(Html.fromHtml(s));
        holder.date.setText(posts.getDate());
        holder.likesCnt.setText(posts.getLikesCnt()+" likes");

        if (posts.isLiked()==true)holder.likeBtn.setImageResource(R.drawable.liked);
        else holder.likeBtn.setImageResource(R.drawable.like);

        holder.likeBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (fragmentButtonListener!=null){
                    if (posts.isLiked()){
                        holder.likeBtn.setImageResource(R.drawable.like);
                        //fragmentButtonListener.myClick(news, 2);
                        fragmentLikeListener.removeItemLike(posts);
                        posts.setLiked(false);
                    } else
                    {
                        holder.likeBtn.setImageResource(R.drawable.liked);
                        fragmentButtonListener.myClick(posts, 1);
                        posts.setLiked(true);
                    }
                }
            }
        });

        holder.itemView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (listener!=null)
                    listener.ItemClick(position, posts);
            }
        });

    }

    @Override
    public int getItemCount() {
        return Posts.postsList.size();
    }

    public class NewsViewHolder extends RecyclerView.ViewHolder{
        TextView author;
        TextView likesCnt;
        TextView postText;
        TextView date;
        ImageView postImage;
        ImageView profilePhoto;
        ImageView likeBtn;
        ImageView saveBtn;

        public NewsViewHolder(@NonNull final View itemView){
            super(itemView);
            author = itemView.findViewById(R.id.author);
            likesCnt = itemView.findViewById(R.id.likesCnt);
            postImage = itemView.findViewById(R.id.postImage);
            postText = itemView.findViewById(R.id.postText);
            date = itemView.findViewById(R.id.date);
            profilePhoto = itemView.findViewById(R.id.profilePhoto);
            likeBtn = itemView.findViewById(R.id.likeBtn);
            saveBtn = itemView.findViewById(R.id.saveBtn);
        }
    }

    public int getItemViewType(int position){
        return position;
    }

    public interface FragmentButtonListener{
        void myClick(Posts posts, int option);
    }

    interface ItemClickListener{
        void ItemClick(int position, Posts item);
    }
    public interface FragmentLikeListener{
        void removeItemLike(Posts posts);
    }

    public void removeLike(Posts posts){
        int n = Posts.postsList.indexOf(posts);
        posts.setLiked(false);
        posts.setLikeBtn(R.drawable.like);
        Posts.postsList.set(n, posts);
        main_list.set(n, posts);
        this.notifyItemChanged(n);
    }
}

File: InstagramLab/app/src/main/java/com/example/instagramlab/PostsAdapter.kt
package com.example.instagramlab


import android.content.Intent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.content.Context
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.example.instagramlab.model.Post
import com.squareup.picasso.Picasso

class PostsAdapter(val posts: ArrayList<Post>, var context: Context): RecyclerView.Adapter<PostsAdapter.ViewHolder>() {

    override fun getItemCount()= posts.size

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view:View = LayoutInflater.from(parent.context).inflate(R.layout.post_row, parent,false)
        return ViewHolder(view)
    }


    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.username.text = posts[position].username
        holder.text.text = posts[position].text
        holder.prosmotri.text = posts[position].prosmotri
        holder.name.text = posts[position].name
        holder.time.text = posts[position].time
        Picasso.get().load(posts[position].photo).into(holder.photo)
        Picasso.get().load(posts[position].avatar).into(holder.avatar)

        holder.detail.setOnClickListener {
            val intent = Intent(context,PostDetailsActivity::class.java)
            intent.putExtra("username",posts[position].username)
            intent.putExtra("text",posts[position].text)
            intent.putExtra("photo",posts[position].photo)
            intent.putExtra("avatar",posts[position].avatar)
            intent.putExtra("prosmotri",posts[position].prosmotri)
            intent.putExtra("name",posts[position].name)
            intent.putExtra("time",posts[position].time)
            context.startActivity(intent)
        }
    }

    class ViewHolder(itemView:View) : RecyclerView.ViewHolder(itemView) {
        val username:TextView = itemView.findViewById(R.id.username)
        val text: TextView = itemView.findViewById(R.id.text)
        val photo : ImageView = itemView.findViewById(R.id.photo)
        val avatar: ImageView = itemView.findViewById(R.id.avatar)
        val prosmotri:TextView = itemView.findViewById(R.id.prosmotri)
        val name:TextView=itemView.findViewById(R.id.name)
        val time:TextView=itemView.findViewById(R.id.time)
        val detail:TextView = itemView.findViewById(R.id.detail)

    }

}

